src/app/archive/archive_lib/processor.ml

```diff
@@ -1,264 +1,539 @@
+    Conn.find_opt
+      (Caqti_request.find_opt Caqti_type.string Caqti_type.int
+         "SELECT id FROM blocks WHERE state_hash = ?")
+      (State_hash.to_string state_hash)
 
-  let create hasura_endpoint = {hasura_endpoint}
+  let load (module Conn : CONNECTION) ~(id : int) =
+    Conn.find
+      (Caqti_request.find Caqti_type.int typ
+         "SELECT state_hash, parent_id, creator_id, snarked_ledger_hash_id, \
+          ledger_hash, height, timestamp, coinbase_id FROM blocks WHERE id = ?")
+      id
 
-  let run t reader =
-    Strict_pipe.Reader.iter reader ~f:(function
-      | Diff.Transition_frontier
-          (Breadcrumb_added {block; sender_receipt_chains_from_parent_ledger})
-        -> (
-          match%bind
-            added_transition t block
-              (Public_key.Compressed.Map.of_alist_exn
-                 sender_receipt_chains_from_parent_ledger)
-          with
-          | Error e ->
-              Graphql_lib.Client.Connection_error.ok_exn e
-          | Ok () ->
-              Deferred.return () )
-      | Diff.Transition_frontier _ ->
-          (* TODO: Implement *)
-          Deferred.return ()
-      | Transaction_pool {added; removed= _} -> (
-          match%bind
-            added_transactions t @@ User_command.Map.of_alist_exn added
-          with
-          | Ok result ->
-              Deferred.return result
-          | Error e ->
-              Graphql_lib.Client.Connection_error.ok_exn e ) )
+  let add_if_doesn't_exist (module Conn : CONNECTION)
+      ({data= t; hash} : (External_transition.t, State_hash.t) With_hash.t) =
+    let open Deferred.Result.Let_syntax in
+    match%bind find (module Conn) ~state_hash:hash with
+    | Some block_id ->
+        return block_id
+    | None ->
+        let%bind parent_id =
+          find (module Conn) ~state_hash:(External_transition.parent_hash t)
+        in
+        let%bind creator_id =
+          Public_key.add_if_doesn't_exist
+            (module Conn)
+            (External_transition.block_producer t)
+        in
+        let%bind snarked_ledger_hash_id =
+          Snarked_ledger_hash.add_if_doesn't_exist
+            (module Conn)
+            ( External_transition.blockchain_state t
+            |> Blockchain_state.snarked_ledger_hash )
+        in
+        let%bind block_id =
+          Conn.find
+            (Caqti_request.find typ Caqti_type.int
+               "INSERT INTO blocks (state_hash, parent_id, creator_id, \
+                snarked_ledger_hash_id, ledger_hash, height, timestamp, \
+                coinbase_id) VALUES (?, ?, ?, ?, ?, ?, ?, ?) RETURNING id")
+            { state_hash= hash |> State_hash.to_string
+            ; parent_id
+            ; creator_id
+            ; snarked_ledger_hash_id
+            ; ledger_hash=
+                External_transition.blockchain_state t
+                |> Blockchain_state.staged_ledger_hash
+                |> Staged_ledger_hash.ledger_hash |> Ledger_hash.to_string
```

Comment:
I remember when we discuss the design of the new schema, Nathan suggested that `snarked_ledger_hash` should have it's own table because we might want to query the archive node by snarked_ledger_hash.